pkg/storage/unified/sql/db/otel/otel_test.go

```diff
@@ -0,0 +1,396 @@
+package otel
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"testing"
+
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/otel/attribute"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	sdktracetest "go.opentelemetry.io/otel/sdk/trace/tracetest"
+	"go.opentelemetry.io/otel/trace"
+
+	"github.com/grafana/grafana/pkg/storage/unified/sql/db"
+	dbmocks "github.com/grafana/grafana/pkg/storage/unified/sql/db/mocks"
+	"github.com/grafana/grafana/pkg/util/testutil"
+)
+
+var errTest = errors.New("because of reasons")
+
+func TestContextAttributes(t *testing.T) {
+	t.Parallel()
+
+	// test context. Note that it's perfectly safe to use context.Background()
+	// and there is no risk of the test blocking at any point because we will
+	// not use the deadline or signal cancellation features of the context
+	ctx := context.Background()
+
+	// test attributes
+	attr1 := attribute.String("the key", "the value")
+	attr2 := attribute.String("the other key", "the other value")
+	attr3 := attribute.String("why not", "have another value")
+	attr4 := attribute.String("it's free", "they say")
+
+	// the subtests are not Parallel because we define this test as a storyline,
+	// since we are interested in testing state changes in the context
+
+	t.Run("consumeAttributes returns nil if SetAttributes was never called",
+		func(t *testing.T) {
+			attrs := consumeAttributes(ctx)
+			require.Nil(t, attrs)
+		})
+
+	t.Run("setting and getting attributes", func(t *testing.T) {
+		ctx = SetAttributes(ctx, attr1, attr2)
+		attrs := consumeAttributes(ctx)
+		require.Len(t, attrs, 2)
+		require.Equal(t, attr1, attrs[0])
+		require.Equal(t, attr2, attrs[1])
+	})
+
+	t.Run("attributes are now cleared", func(t *testing.T) {
+		attrs := consumeAttributes(ctx)
+		require.Len(t, attrs, 0)
+	})
+
+	t.Run("SetAttributes overwrites previous attributes", func(t *testing.T) {
+		ctx = SetAttributes(ctx, attr1, attr2)
+		ctx = SetAttributes(ctx, attr3, attr4)
+		attrs := consumeAttributes(ctx)
+		require.Len(t, attrs, 2)
+		require.Equal(t, attr3, attrs[0])
+		require.Equal(t, attr4, attrs[1])
+	})
+
+	t.Run("attributes are now cleared again", func(t *testing.T) {
+		attrs := consumeAttributes(ctx)
+		require.Len(t, attrs, 0)
+	})
+}
+
+func TestOTelTransactions(t *testing.T) {
```

Comment:
This test is exercising the exact model in the example [otel.go](https://github.com/grafana/grafana/pull/92266/files#diff-8b107852ef49e18c4fed77d10202bb12bccb056e527885ac4370d43cfce91d16R266-R281).